src/bp/core/services/retention/service.ts

```diff
@@ -0,0 +1,105 @@
+import { RetentionPolicy } from 'core/config/botpress.config'
+import { ConfigProvider } from 'core/config/config-loader'
+import Database from 'core/database'
+import { TYPES } from 'core/types'
+import diff from 'deep-diff'
+import { inject, injectable } from 'inversify'
+import _ from 'lodash'
+import moment from 'moment'
+import ms from 'ms'
+
+import { getPaths } from './util'
+
+export interface ExpiredData {
+  channel: string
+  user_id: string
+  field_path: string
+}
+
+@injectable()
+export class DataRetentionService {
+  private readonly tableName = 'data_retention'
+  private policies: RetentionPolicy[] | undefined
+  private DELETED_ATTR = 'D'
+
+  constructor(
+    @inject(TYPES.ConfigProvider) private configProvider: ConfigProvider,
+    @inject(TYPES.Database) private database: Database
+  ) {}
+
+  async initialize(): Promise<void> {
+    const config = await this.configProvider.getBotpressConfig()
+    this.policies = config.dataRetention && config.dataRetention.policies
+  }
+
+  async hasPolicy() {
+    return this.policies
+  }
+
+  async checkChanges(channel: string, user_id: string, beforeAttributes: any, afterAttributes: any) {
+    const differences = diff(getPaths(beforeAttributes), getPaths(afterAttributes))
+    if (!differences || !this.policies) {
+      return
+    }
+
+    const changedPaths = _.flatten(differences.filter(diff => diff.kind != this.DELETED_ATTR).map(diff => diff.path))
+    if (!changedPaths.length) {
+      return
+    }
+
+    for (const field in this.policies) {
+      if (changedPaths.indexOf(field) > -1) {
+        const expiry = moment()
+          .add(ms(this.policies[field]), 'ms')
+          .toDate()
+
+        if (await this.get(channel, user_id, field)) {
+          await this.update(channel, user_id, field, expiry)
```

Comment:
we have an upsert knex extension for this, because doing it in two async calls lead to potential race conditions (i.e. .get succeeds but then expires before .update is called)